File: mergeTwoLists.cpp

/* @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>

using namespace std;

//#define testMod

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod

typedef struct {
    int val;
    ListNode *next;
    //ListNode(int x) : val(x), next(NULL) {}
}ListNode ;

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *end=NULL,*out=NULL;
        
        ListNode *tmp1=l1,*tmp2=l2;
        if(l1==NULL)
            return l2;
        else if(l2==NULL)
            return l1;
        if(tmp1->val< tmp2->val)
        {
            end=tmp1;
            out=tmp1;
            tmp1=tmp1->next;
        }
        else
        {
            end=tmp2;
            out=tmp2;
            tmp2=tmp2->next;
        }
        while(tmp1!=NULL&&tmp2!=NULL)
        {
            if(tmp1->val < tmp2->val)
            {
                end->next=tmp1;
                end=end->next;
                tmp1=tmp1->next;
            }
            else
            {
                end->next=tmp2;
                end=end->next;
                tmp2=tmp2->next;
            }
        }
        if(tmp1!=NULL)
            end->next=tmp1;
        else
            end->next=tmp2;

        return out;
    }
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;

#endif 
    return 0;
}

File: mergeTwoLists.py
from typing import Optional

from pylist import ListNode


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode],
                      list2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode(-1)
        p = res

        while list1 is not None and list2 is not None:
            if list1.val < list2.val:
                p.next = list1
                p = list1
                list1 = list1.next
            else:
                p.next = list2
                p = list2
                list2 = list2.next

        p.next = list1

        while p.next is not None:
            p = p.next
        p.next = list2

        return res.next


l1 = [1, 2, 4]
l2 = [1, 3, 4]
l1 = []
l2 = []
l1 = []
l2 = [0]
l1 = ListNode.fromList(l1)
l2 = ListNode.fromList(l2)
ListNode.travel(Solution().mergeTwoLists(l1, l2))
